### PR TITLE
enforce only one copy of replica

### PR DESCRIPTION
**Pull request by tbg**
- LocalSender will error out if it has a range in two stores
  (with a lot of stores, maybe we want to optimize a little
  bit the fact that now it is doing a linear search for each
  request, but that is an isolated improvement we can do when
  ever we need it)
- ChangeReplicas will also error out when asked to put a replica
  on the same node with one already present.
- the allocator seems node-aware already, so that part looked ok.
- introduced a multiLocalSender for the multiTestContext, which
  adds any additional stores to isolated local senders. This
  avoids three test failures in storage/client_raft_test related
  to replication and is cleaner anyways, since the new stores
  are made believe to belong to different nodes.

### CODE REVIEW

File: storage/client_test.go [hunk 1]
```diff
@@ -96,12 +96,31 @@ func createTestStoreWithEngine(t *testing.T, eng engine.Engine, clock *hlc.Clock
...
+	mls.senders[index] = s
+}
+
 type multiTestContext struct {
 	t           *testing.T
 	manualClock *hlc.ManualClock
 	clock       *hlc.Clock
 	gossip      *gossip.Gossip
 	sender      *kv.LocalSender
+	multisender *multiLocalSender
```
Comment (spencerkimball) [inline storage/client_test.go]: Maybe just a slice like for engines would be simpler.

**Comment by spencerkimball** [comment]
LGTM

File: storage/client_test.go [hunk 2]
```diff
@@ -139,14 +158,19 @@ func (m *multiTestContext) Start(t *testing.T, numStores int) {
 	if m.transport == nil {
 		m.transport = multiraft.NewLocalRPCTransport()
 	}
+
```
Comment (spencerkimball) [inline storage/client_test.go]: Why do we still have the sender, btw?

File: storage/client_test.go (hunk 1 above)
Comment (tschottdorf) [inline storage/client_test.go]: done

File: kv/local_sender.go [hunk 3]
```diff
@@ -142,13 +142,28 @@ func (ls *LocalSender) Send(call client.Call) {
...
+		if rng == nil {
+			continue
 		}
+		if replica == nil {
+			raftID = rng.Desc().RaftID
+			replica = rng.GetReplica()
+			continue
+		}
+		// Should never happen outside of tests.
+		return 0, nil, util.Errorf(
```
Comment (petermattis) [inline kv/local_sender.go]: Checking for multiple replicas on the same store during lookup feels wrong. Shouldn't we be performing this checking when a range is added to a store?

File: kv/local_sender.go [hunk 4]
```diff
@@ -142,13 +142,28 @@ func (ls *LocalSender) Send(call client.Call) {
 // by consulting each store in turn via Store.LookupRange(key).
 // Returns RaftID and replica on success; RangeKeyMismatch error
 // if not found.
-func (ls *LocalSender) lookupReplica(start, end proto.Key) (int64, *proto.Replica, error) {
+// TODO(tschottdorf) with a very large number of stores, the LocalSender
+// may want to avoid scanning the whole map of stores on each invocation.
```
Comment (petermattis) [inline kv/local_sender.go]: Seems like `LocalSender` (and perhaps `Store`) should be using something like `kv.RangeCache` where an interval tree from key range to range is kept.

File: kv/local_sender.go (hunk 3 above)
Comment (tschottdorf) [inline kv/local_sender.go]: this is a sanity check as there were tests that violated this. The replication queue is making sure to never put more than one copy of each replica on a node, and the store.AddReplica checks it again (see this PR).

Comment (petermattis) [inline kv/local_sender.go]: Ah, you are checking in `ChangeReplicas`. So why are you also checking here? Paranoia?

Comment (tschottdorf) [inline kv/local_sender.go]: healthy dose of paranoia plus making sure that we don't write tests that set up multiple replicas on one store manually.

**Pull request merged**